src/extractor/extractor_callbacks.cpp

```diff
@@ -175,6 +175,38 @@ void ExtractorCallbacks::ProcessWay(const osmium::Way &input_way, const Extracti
         }
     }
 
+    const auto classStringToMask = [this](const std::string &class_name) {
+        auto iter = classes_map.find(class_name);
+        if (iter == classes_map.end())
+        {
+            if (classes_map.size() > MAX_CLASS_INDEX)
+            {
+                throw util::exception("Maximum number of classes if " +
+                                      std::to_string(MAX_CLASS_INDEX + 1));
+            }
+            ClassData class_mask = 1u << classes_map.size();
+            classes_map[class_name] = class_mask;
+            return class_mask;
+        }
+        else
+        {
+            return iter->second;
+        }
+    };
+    const auto classesToMask = [&](const auto &classes) {
+        ClassData mask = 0;
+        for (const auto &pair : classes)
+        {
+            if (pair.second)
+            {
+                mask |= classStringToMask(pair.first);
+            }
+        }
+        return mask;
+    };
```

Comment:
Hm this is a bit hard to follow with the auto and first/second working on basically arbitrary types - can you make it a bit more explicit?